ops/main.py

```diff
@@ -42,6 +43,114 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
+class _JujuContext:
+    # source: https://juju.is/docs/juju/charm-environment-variables
+    # the HookVars function: https://github.com/juju/juju/blob/3.6/worker/uniter/runner/context/context.go#L1398
+    # only allow accesing these vars because these are the currently used vars in main.py
+
+    # env var JUJU_ACTION_NAME
+    action_name: Optional[str] = None
+
+    # env var JUJU_ACTION_UUID
+    action_uuid: Optional[str] = None
+
+    # env var JUJU_CHARM_DIR
+    charm_dir: Optional[str] = None
+
+    # env var JUJU_DEBUG
+    debug: Optional[str] = None
+
+    # env var JUJU_DEPARTING_UNIT
+    departing_unit: Optional[str] = None
+
+    # env var JUJU_DISPATCH_PATH
+    dispatch_path: Optional[str] = None
+
+    # env var JUJU_MODEL_NAME
+    model_name: Optional[str] = None
+
+    # env var JUJU_MODEL_UUID
+    model_uuid: Optional[str] = None
+
+    # env var JUJU_NOTICE_ID
+    notice_id: Optional[str] = None
+
+    # env var JUJU_NOTICE_KEY
+    notice_key: Optional[str] = None
+
+    # env var JUJU_NOTICE_TYPE
+    notice_type: Optional[str] = None
+
+    # env var JUJU_PEBBLE_CHECK_NAME
+    pebble_check_name: Optional[str] = None
+
+    # env var JUJU_RELATION
+    relation_name: Optional[str] = None
+
+    # env var JUJU_RELATION_ID
+    relation_id: Optional[int] = None
+
+    # env var JUJU_REMOTE_APP
+    remote_app: Optional[str] = None
+
+    # env var JUJU_REMOTE_UNIT
+    remote_unit: Optional[str] = None
+
+    # env var JUJU_SECRET_ID
+    secret_id: Optional[str] = None
+
+    # env var JUJU_SECRET_LABEL
+    secret_label: Optional[str] = None
+
+    # env var JUJU_SECRET_REVISION
+    secret_revision: Optional[int] = None
+
+    # env var JUJU_STORAGE_ID
+    storage_id: Optional[str] = None
+
+    # env var JUJU_UNIT_NAME
+    unit_name: Optional[str] = None
+
+    # env var JUJU_VERSION
+    version: Optional[str] = None
+
+    # env var JUJU_WORKLOAD_NAME
+    workload_name: Optional[str] = None
+
+    @classmethod
+    def from_environ(cls, env: Dict[str, Any]) -> '_JujuContext':
+        return _JujuContext(
+            action_name=env.get('JUJU_ACTION_NAME'),
+            action_uuid=env.get('JUJU_ACTION_UUID'),
+            charm_dir=env.get('JUJU_CHARM_DIR'),
+            debug=env.get('JUJU_DEBUG'),
+            departing_unit=env.get('JUJU_DEPARTING_UNIT'),
+            dispatch_path=env.get('JUJU_DISPATCH_PATH'),
+            model_name=env.get('JUJU_MODEL_NAME'),
+            model_uuid=env.get('JUJU_MODEL_UUID'),
+            notice_id=env.get('JUJU_NOTICE_ID'),
+            notice_key=env.get('JUJU_NOTICE_KEY'),
+            notice_type=env.get('JUJU_NOTICE_TYPE'),
+            pebble_check_name=env.get('JUJU_PEBBLE_CHECK_NAME'),
+            relation_name=env.get('JUJU_RELATION'),
+            relation_id=int(env.get('JUJU_RELATION_ID', '').split(':')[-1])
+            if 'JUJU_RELATION_ID' in env
+            else None,
```

Comment:
Can we add some parentheses to improve the formatting here? Also, similar to above, might be best to use `if env.get('JUJU_RELATION_ID')` in case it's set but empty string.
```suggestion
            relation_id=(int(env['JUJU_RELATION_ID'].split(':')[-1])
                         if env.get('JUJU_RELATION_ID')
                         else None),
```